### PR TITLE
[WIP] Rewrote adam optimizer with foreach APIs

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43926 Rewrote Adadelta with _foreach APIs
* #43925 Rewrote adamax optimizer with _foreach APIs
* #43727 Rewrote asgd optimizer with *_foreach APIs
* #43726 Rewrote rprop optimizer with _foreach APIs
* #43725 Rewrote rmsprop with _forerach APIs
* #43724 Rewrote SGD with _foreach APIs
* #43562 Rewrote AdamW with _foreach APIs
* **#43507 [WIP] Rewrote adam optimizer with foreach APIs**

Differential Revision: [D23331893](https://our.internmc.facebook.com/intern/diff/D23331893)

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same dtype.
- All Tensors must be on the same device. 

----------------
**In this PR**
- We are introducing new namespace under torch.optim - torch.optim.multi_tensor, where we will have optimizers rewritten with _foreach_* APIs.
- Rewriting adam optimizer with _foreach_* APIs

### CODE REVIEW

File: torch/optim/multi_tensor/adam.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+                state['step'] += 1
+                states.append(state)
+
+            beta1, beta2 = group['betas']
+
+            bias_correction1 = [1 - beta1 ** state['step'] for state in states] 
+            bias_correction2 = [1 - beta2 ** state['step'] for state in states] 
+            if group['weight_decay'] != 0:
+                torch._foreach_add_(grads, group['params'], group['weight_decay'])
```
Comment (ngimel) [inline torch/optim/multi_tensor/adam.py:115]: In the original optimizer the line is 
```
grad = grad.add(p, alpha=group['weight_decay'])
```
This is not inplace, and original grad attributes (p.grad) aren't mutated. Here you are mutating `p.grad `inplace (I'm surprised tests don't catch it).

File: torch/optim/multi_tensor/adam.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            #
+            torch._foreach_mul_(exp_avg, beta1)
+            torch._foreach_add_(exp_avg, grads, 1 - beta1)
+
+            torch._foreach_mul_(exp_avg_sq, beta2)
+            torch._foreach_addcmul_(exp_avg_sq, grads, grads, 1 - beta2)
+
+            if amsgrad:
+                # Maintains the maximum of all 2nd moment running avg. till now
+                max_exp_avg_sq = [torch.max(a, b) for a, b in zip(max_exp_avg_sq, exp_avg_sq)]
```
Comment (ngimel) [inline torch/optim/multi_tensor/adam.py:128]: Aha, so ideally we also need a foreach max, because now it will be a loop?

File: torch/optim/multi_tensor/adam.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                # Use the max. for normalizing running avg. of gradient
+                max_exp_avg_sq_sqrt = torch._foreach_sqrt(max_exp_avg_sq)
+                bias_correction_sqrt = [math.sqrt(bc) for bc in bias_correction2]
+                max_exp_avg_sq_sqrt = [torch.div(a, b) for a, b in zip(max_exp_avg_sq_sqrt, bias_correction_sqrt)]
```
Comment (ngimel) [inline torch/optim/multi_tensor/adam.py:132]: Ok, so this is a loop because we don't have Op(TensorList, ScalarList)? This is unfortunate, looks like we really need it.

File: torch/optim/multi_tensor/adam.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                exp_avg_sq_sqrt = torch._foreach_sqrt(exp_avg_sq)
+                bias_correction_sqrt = [math.sqrt(bc) for bc in bias_correction2]
+
+                exp_avg_sq_sqrt = [torch.div(a, b) for a, b in zip(exp_avg_sq_sqrt, bias_correction_sqrt)]
+                denom = torch._foreach_add(exp_avg_sq_sqrt, group['eps'])
+
+            step_size = [group['lr'] / bc for bc in bias_correction1]
+
+            for i in range(len(step_size)):
+                params_with_grad[i].addcdiv_(exp_avg[i], denom[i], value=-step_size[i])
```
Comment (ngimel) [inline torch/optim/multi_tensor/adam.py:144]: And this is another case where we need op(TensorList, ScalarList), because params_with_grad is a tensorlist, exp_avg is a tensorlist, denom is a TensorList, and value is ScalarList?

**Review by ngimel — COMMENTED** [review]
This generally looks good, but it reveals that we need a few more (TensorList,  ScalarList) operations, because now there are still a few loops over parameters/grads.

File: torch/optim/multi_tensor/adam.py (hunk 3 above)
Comment (ngimel) [reply torch/optim/multi_tensor/adam.py:132]: Per conversation in slack, we can check if all `step`s are the same, that will guarantee that all bias_corrections are the same, and we can use for_each here and in other places if it's the case (should be common).

File: torch/optim/multi_tensor/adam.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+            torch._foreach_mul_(exp_avg_sq, beta2)
+            torch._foreach_addcmul_(exp_avg_sq, grads, grads, 1 - beta2)
+
+            if amsgrad:
+                # Maintains the maximum of all 2nd moment running avg. till now
+                max_exp_avg_sq = [torch.max(a, b) for a, b in zip(max_exp_avg_sq, exp_avg_sq)]
+                # Use the max. for normalizing running avg. of gradient
+                max_exp_avg_sq_sqrt = torch._foreach_sqrt(max_exp_avg_sq)
+                bias_correction_sqrt = [math.sqrt(bc) for bc in bias_correction2]
```
Comment (wanchaol) [inline torch/optim/multi_tensor/adam.py:131]: why this is not a for_each_sqrt?

Comment (izdeby) [reply torch/optim/multi_tensor/adam.py:131]: in this context, bias_correction2 is a list of scalars. there is no foreach api that supports lists of scalars. im working on those APIs right now.

Comment (ngimel) [reply torch/optim/multi_tensor/adam.py:131]: This particular case is ok, those are python scalars and this operation is reasonably fast. However, other cases where cuda tensors are interacting with a list of scalars are more problematic. We could get around them for now by checking if all bias_corrections have the same value (which should be a common case).

Comment (izdeby) [reply torch/optim/multi_tensor/adam.py:131]: I gave myself 2 days to fight codegen and make ScalarList a reality. if it works out - i will just add new APIs for `_foreach_op(TensorList, ScalarList)`. And if it will be too complex, will make a workaround and add this to TODO.

**Comment by vincentqb** [comment]
- What would be the advantage of using the previous version? One noted in the description is that the dtype has to be the same. (We did mention offline that MultiTensor would not be differentiable for instance.)
- Can TensorList be reduced to list of tensors if a list of tensors is provided instead? If so could we consider having only this new implementation for each optimizer?
- Should we consider having a wrapper that swaps between the two current implementation depending on the type? 

> Broadcasting
> At this point we don't support broadcasting.

This mean some operations could be slower with MultiTensor for now though, right?

**Comment by vincentqb** [comment]
> [GitHub issue](https://github.com/pytorch/pytorch/issues/38655)
> Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start.
> As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex).
> In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

Now that we have the implementation, have you been able to quantify the performance scaling?

File: torch/optim/multi_tensor/adam.py (hunk 3 above)
Comment (vincentqb) [reply torch/optim/multi_tensor/adam.py:132]: Are you suggesting a change in the underlying algorithm? What does it mean and what would we happens if all steps are not the same?

File: test/test_optim.py [hunk 6]
```diff
@@ -294,47 +295,48 @@ def test_sgd_sparse(self):
...
-                lr=1e-3, amsgrad=True),
-            [lambda opt: StepLR(opt, gamma=0.9, step_size=10),
-             lambda opt: ReduceLROnPlateau(opt)]
-        )
-        with self.assertRaisesRegex(ValueError, "Invalid beta parameter at index 0: 1.0"):
-            optim.Adam(None, lr=1e-2, betas=(1.0, 0.0))
-
-        with self.assertRaisesRegex(ValueError, "Invalid weight_decay value: -1"):
-            optim.Adam(None, lr=1e-2, weight_decay=-1)
+        for optimizer in [optim.Adam, optim_mt.Adam]:
```
Comment (vincentqb) [inline test/test_optim.py:298]: note: current tests don't guarantee that the algorithm is the same as the other, or even that there's convergence in either case.

@zou3519 @anjali411 -- how are the C++ APIs tested?

Comment (anjali411) [reply test/test_optim.py:298]: C++ Optimizer API logic is tested here: https://github.com/pytorch/pytorch/blob/master/test/cpp/api/optim.cpp#L311. These tests compare the C++ optimizers' results to the Python API optimizers' results prewritten in this file: https://github.com/pytorch/pytorch/blob/master/test/cpp/api/optim_baseline.h

Comment (ngimel) [reply test/test_optim.py:298]: Ok, so here we'll need the tests comparing optim.Adam results with optim_mt.Adam, in addition to `_test_basic_cases`?

Comment (vincentqb) [reply test/test_optim.py:298]: Yes, since the contract here is that the MultiTensor implementation is the same as the original.

Comment (vincentqb) [reply test/test_optim.py:298]: Is there also a C++ equivalent with the foreach API for the C++ optimizer?

Comment (izdeby) [reply test/test_optim.py:298]: @vincentqb, there will be C++ optimizers as well but a bit later. We decided to start with python ones first. 

re testing: Is there anything specific you would suggest testing?

Comment (vincentqb) [reply test/test_optim.py:326]: Ideally, we'd have a test as mentioned by @anjali411 [above](https://github.com/pytorch/pytorch/pull/43507#discussion_r486502986) that checks that the two implementations give the exact same answer in some case.

Note that if this implementation were directly replacing the current one, the C++ test would also tell us that the implementations are still aligned :) Maybe there's a way of leveraging those tests already there?

Comment (ngimel) [reply test/test_optim.py:326]: There are tests in apex comparing optimizer implementations, they can be adopted here if c++ tests are hard to use for some reason https://github.com/NVIDIA/apex/blob/master/tests/L0/run_optimizers/test_fused_optimizer.py

**Comment by stas00** [comment]
this is wonderful! thank you!

I don't see any user docs discussing this improvement.

What is the implication for the user?

Should we switch to use `torch.optim.multi_tensor` if the version is right, or is the intention for it to eventually take over `torch.optim` and no changes need to be done to the user code.

Thank you.

**Comment by izdeby** [comment]
Hi, @stas00 
First of all, if you would like to track the progress of this work, please take a look at [this stack](https://github.com/pytorch/pytorch/pull/48223).

Answering your question, you can try using the optimizers from `torch.optim._multi_tensor` namespace, but please keep in mind that they are in an alpha state and will be updated. Eventually, they will replace the ones from `torch.optim`. Once that happens, users dont have to do anything at all. The changes will happen under the hood.

**Comment by stas00** [comment]
Thank you very much for the clarification and the stack, @izdeby!

So basically we have an option to deploy these early for those who need the speed up sooner, but otherwise there is nothing to be done.

Excellent!

**Comment by Wesley-Jzy** [comment]
If I want to implement global operators like grad_clip to reduce kernel launch by myself, may I use multi_tensor to do it or PyTorch just provide similar interfaces?

**Comment by ngimel** [comment]
Pytorch's `clip_grad_norm` uses multi_tensor, if you want to write your own utility you can use multi_tensor yourself.

**Comment by Wesley-Jzy** [comment]
> Pytorch's `clip_grad_norm` uses multi_tensor, if you want to write your own utility you can use multi_tensor yourself.

Thank you so much. I think multi_tensor is what I'm looking for.
